### PR TITLE
support named groups

### PR DESCRIPTION
**Pull request by nathanrpage97**
Allow named groups flag `"n"` to be available

### CODE REVIEW

**Comment by kmarius** [comment]
I think it would be a good idea to also store a copy of the named group at the corresponding index. With your current solution, `ipairs` won't work and `pairs` doesn't respect insertion order. Then again, keeping named groups in the same table seems a bit awkward. It seems in javascript, indexed groups are accessed by indexing the match result (i.e. stored in the top level "table") and the named groups accessed via the `groups` field.

**Comment by nathanrpage97** [comment]
How about a separate `named_groups` table is created and the results are duplicated?

File: jsregexp.c [hunk 1]
```diff
@@ -63,13 +64,33 @@ static int regex_closure(lua_State *lstate)
...
+      } 
+    }
+    if (named_groups) {
+      lua_setfield(lstate, -3, "named_groups");
+    } else {
+      lua_pop(lstate, 1);
     }
     lua_setfield(lstate, -2, "groups");
-
+
```
Comment (kmarius) [inline jsregexp.c]: whitespace

File: jsregexp.c [hunk 2]
```diff
@@ -63,13 +64,33 @@ static int regex_closure(lua_State *lstate)
     lua_pushinteger(lstate, capture[1] - input);
     lua_setfield(lstate, -2, "end_ind");
 
-    lua_newtable(lstate);
+    lua_newtable(lstate); // groups
+    lua_newtable(lstate); // named_groups
+
+    const char* group_names = NULL;
+    if (named_groups) {
+      group_names = lre_get_groupnames(r->bc);
```
Comment (kmarius) [inline jsregexp.c:72]: I think we can just  push the table for named groups here so we don't have to pop it later. Need to switch the stack indexes `-2` and `-3` in the loop, of course.

Comment (nathanrpage97) [reply jsregexp.c:72]: Uhh I’m not quite sure how to do this without having some sort of ternarys  to specify stack position based on if named groups is true

Comment (kmarius) [reply jsregexp.c:72]: I should have made this clearer. If the table of the normal indexed matches is on top, one can always access it with `rawseti(lstate, -2, ...)`. If `named_groups` is true, have the table of named groups *below* that so it is accessed with `setfield(lstate, -3, ...)`. Just make sure you push a second table when `named_groups` is true.

Comment (nathanrpage97) [reply jsregexp.c:72]: I have updated it to work somewhat as you describe. I'm pretty sure you have to do some branching for stack positions as is done [here](https://github.com/kmarius/jsregexp/pull/8/files#diff-bf832cc2ca0e905b0def25d6989e2ead4b330c512a0a651094d8eaeee86cc4e8R88-R92)

**Comment by kmarius** [comment]
I went ahead and merged because I thought it would be easier merging before #5. Thanks for the contribution!

**Pull request merged**